### ISSUE TITLE
Consider spliting RUIProvider and exporting HOCs

### ISSUE THREAD

**Issue by mbohal** [react-ui-org/react-ui#579]
Presently we use https://github.com/react-ui-org/react-ui/blob/master/src/provider/RUIContext.jsx to pass down to components both [globalProps](https://react-ui.io/docs/customize/global-props/) and [translations](https://react-ui.io/docs/customize/translations/). I think we should separate these into two independent contexts.

**Reasons:**

1. They have nothing in common, even the documentation is split into two pages
2. When implementing `globalProps` in a project it would be useful to be able to extend the `globalProps` system to other project specific React components (e.g. some custom `PasswordField`).
3. It might improve performance, because dynamically  changing `globalProps` will not rerender components that from `RUIProvider` only consume the  `translations` object and vice versa. 

**Export:**

If we decide to support using `globalProps` and perhaps even `translations` to custom components, we would also need to:

* export `withGlobalProps`
* create, use for RUI components and export `withTranslations`
* update docs

**Comment by bedrich-schindler** [comment]
I agree with splitting`RUIContext` into`RUIGlobalPropsContext` and `RUITranslationsContext`, but ...

I strongly disagree with the rest. The whole concept of `RUIContext` is to configure React UI. If you use React UI to implement your UI system, it is up to you to wrap it and solve it at your UI system. RUI context is not place to save your own data. I do not like idea interconnecting RUI, lets wrap it in React way. Not interconnect it with your own data. We are not here to solve every problem.

If anyone else have different opinion or see advantage of this, lets discuss it. The only think I would consider is to pass you own Context to the RUIGlobalPropsContext` and `RUITranslationsContext`. This would allow you to simplify hydrating RUI with the data you want and at the same time you do not need to have two contexts in your UI system. But exporting internal hocs might introduce more problems in the future as every change in this internal function would be consider breaking-change.

**Comment by mbohal** [comment]
Just so we all agree what we are talking about, this was my idea:

```jsx
<RUIProvider
  globalProps={{
    TextField: { layout: 'horizontal' },
    MyField: { layout: 'horizontal' },
  }}
>
  <TextField … />
  <MyField … />
</RUIProvider>
```

The only thing that changes is the shape of the object in `globalProps`.

@bedrich-schindler  When you wrote:

>  The only think I would consider is to pass you own Context to the RUIGlobalPropsContextandRUITranslationsContext

is this what you had in mind?

---

>  If you use React UI to implement your UI system, it is up to you to wrap it and solve it at your UI system

I think I disagree with this because:

1. RUI is primarly aimed at providing a convenient base for building your own UI system.
2. Building a UI system typically involves adding components on top of the base components provided by RUI
3. So it is within the scope of RUI to provide utility tools to make extension easy (`transferprops`, `className`, etc.).

I think using RUI as a base for your own UI system is the primary usecase. Using it directly I see as supported, but secondary.

**Comment by mbohal** [comment]
On call we agreed that @mbohal will add here a sumarization of the features history, to make sure, that we do not run in circles.

**Comment by bedrich-schindler** [comment]
I just want to point out that we did change from the desired state of this PR in https://github.com/react-ui-org/react-ui/pull/365

**Comment by adamkudrna** [comment]
Resolved in #583.
